### PR TITLE
feat(BBD-822): Store the record index for each product

### PR DESCRIPTION
**Pull request by Vic-Dev**

### CODE REVIEW

File: src/core/adapters/search.ts [hunk 1]
```diff
@@ -144,10 +144,26 @@ namespace Adapter {
...
+    return results.records.map(({ allMeta }) => {
+      const meta: any = {};
+
+      if (allMeta[state.session.config.recommendations.idField] in pastPurchases) {
+        meta.pastPurchase = true;
+      }
+
+      return {
+        meta,
+        index: startIndex++,
```
Comment (effervescentia) [inline src/core/adapters/search.ts]: should be able to use the index from `map()` and add it to `startIndex`
make `startIndex` a `const`

File: src/core/selectors.ts [hunk 2]
```diff
@@ -113,23 +113,13 @@ namespace Selectors {
    * Returns the current products
    */
   export const products = (state: Store.State) =>
-    state.data.present.products;
+    Search.extractData(state.data.present.products);
 
   /**
    * Returns the current products extended with metadata
    */
-  export const productsWithMetadata = (state: Store.State, idField: string) => {
```
Comment (effervescentia) [inline src/core/selectors.ts]: resurrect this selector as `productsWithPastPurchase`

File: src/core/adapters/search.ts [hunk 3]
```diff
@@ -144,10 +144,26 @@ namespace Adapter {
...
+    products.map(({ data }) => data);
 
-  export const extractProduct = ({ allMeta }) => allMeta;
+  export const extractProducts = (state: Store.State, results: Results) => {
+    const pastPurchases = Selectors.pastPurchaseProductsBySku(state);
+    let startIndex = results.pageInfo.recordStart;
+    return results.records.map(({ allMeta }) => {
+      const meta: any = {};
+
+      if (allMeta[state.session.config.recommendations.idField] in pastPurchases) {
```
Comment (effervescentia) [inline src/core/adapters/search.ts]: we should only add this when selecting the data back out to ensure we never get out of sync

File: test/unit/core/adapters/search.ts [hunk 4]
```diff
@@ -335,28 +335,48 @@ suite('Search Adapter', ({ expect, stub }) => {
     });
   });
 
+  describe('extractdata()', () => {
```
Comment (effervescentia) [inline test/unit/core/adapters/search.ts]: `extractData()`

File: src/core/adapters/search.ts (hunk 1 above)
Comment (Vic-Dev) [reply src/core/adapters/search.ts]: what?

File: src/core/adapters/search.ts [hunk 5]
```diff
@@ -144,10 +144,21 @@ namespace Adapter {
...
-    results.records.map(Adapter.extractProduct);
-
-  export const extractProduct = ({ allMeta }) => allMeta;
+  export const extractData = (products: Store.ProductWithMetadata[]) =>
+    products.map(({ data }) => data);
+
+  export const augmentProducts = (results: Results) => {
+    const startIndex = results.pageInfo.recordStart;
+    return results.records.map(({ allMeta }, index) => {
+      const meta: any = {};
```
Comment (effervescentia) [inline src/core/adapters/search.ts]: could probably just inline this if we don't plan on mutating it

Comment (effervescentia) [reply src/core/adapters/search.ts]: PS can you extract `collection` from the record and put that in the `meta`?

File: src/core/observer.ts [hunk 6]
```diff
@@ -137,7 +139,10 @@ namespace Observer {
           },
 
           recommendations: {
-            products: emit(Events.RECOMMENDATIONS_PRODUCTS_UPDATED)
+            // products: emit(Events.RECOMMENDATIONS_PRODUCTS_UPDATED)
```
Comment (effervescentia) [inline src/core/observer.ts]: can we remove this comment?

**Pull request merged**